homeassistant/components/modbus/__init__.py

```
@@ -88,163 +97,219 @@
+        if self._config_type == "serial":
+            _, self._client = ClientSerial(
+                schedulers.ASYNC_IO,
+                method=self._config_method,
+                port=self._config_port,
+                baudrate=self._config_baudrate,
+                stopbits=self._config_stopbits,
+                bytesize=self._config_bytesize,
+                parity=self._config_parity,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        elif self._config_type == "rtuovertcp":
+            _, self._client = ClientTCP(
+                schedulers.ASYNC_IO,
+                host=self._config_host,
+                port=self._config_port,
+                framer=ModbusRtuFramer,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        elif self._config_type == "tcp":
+            _, self._client = ClientTCP(
+                schedulers.ASYNC_IO,
+                host=self._config_host,
+                port=self._config_port,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        elif self._config_type == "udp":
+            _, self._client = ClientUDP(
+                schedulers.ASYNC_IO,
+                host=self._config_host,
+                port=self._config_port,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        else:
+            assert False
 
-    def connect(self):
-        """Connect client."""
-        with self._lock:
-            self._client.connect()
+    async def _read(self, unit, address, count, func):
+        """Read generic with error handling."""
+        await self._connect_delay()
+        async with self._lock:
+            kwargs = {"unit": unit} if unit else {}
+            result = await func(address, count, **kwargs)
+            if isinstance(result, (ModbusException, ExceptionResponse)):
```

It would be nice if the client func raised errors instead of returning them. I guess this is the library responsibility.